saleor/graphql/product/bulk_mutations/products.py

```diff
@@ -705,22 +705,54 @@ def perform_mutation(cls, root, info, **data):
             warehouses = cls.get_nodes_or_error(
                 warehouse_ids, "warehouse", only_type=Warehouse
             )
-            cls.update_or_create_variant_stocks(variant, stocks, warehouses)
+            cls.update_or_create_variant_stocks(variant, stocks, warehouses, info)
 
         variant = ChannelContext(node=variant, channel_slug=None)
         return cls(product_variant=variant)
 
     @classmethod
     @traced_atomic_transaction()
-    def update_or_create_variant_stocks(cls, variant, stocks_data, warehouses):
+    def update_or_create_variant_stocks(cls, variant, stocks_data, warehouses, info):
         stocks = []
         for stock_data, warehouse in zip(stocks_data, warehouses):
-            stock, _ = warehouse_models.Stock.objects.get_or_create(
+            stock, created = warehouse_models.Stock.objects.get_or_create(
                 product_variant=variant, warehouse=warehouse
             )
+            stocks.append(
+                {
+                    "stock": stock,
+                    "created": created,
+                    "prev_quantity": stock.quantity,
+                    "current_quantity": stock_data["quantity"],
+                }
+            )
             stock.quantity = stock_data["quantity"]
-            stocks.append(stock)
-        warehouse_models.Stock.objects.bulk_update(stocks, ["quantity"])
+        stocks_to_update = [s["stock"] for s in stocks]
+        warehouse_models.Stock.objects.bulk_update(stocks_to_update, ["quantity"])
+        cls._run_product_variant_stock_exists_webhook(info, stocks)
+        cls._run_product_variant_stock_changed_webhook(info, stocks_to_update)
+
+    @classmethod
+    def _run_product_variant_stock_exists_webhook(cls, info, updated_stocks):
+        for stock in (
+            s["stock"]
+            for s in updated_stocks
+            if cls._product_variant_webhook_conditions(s)
```

Comment:
please don't nest `for` loops, extract the inner part outside.

Reply:
The `_product_variant_webhook_conditions` return not loop but `bool`.

Reply:
I've just meant to get the for out of for :) but maybe it's better this way to keep it compact.

Reply:
I agree with @tomaszszymanski129, nested for-loops make it harder to read.